Sources/OpenTelemetrySdk/Trace/SpanProcessors/BatchSpanProcessor.swift

```diff
@@ -57,40 +72,96 @@ public struct BatchSpanProcessor: SpanProcessor {
+        meterProvider: StableMeterProvider,
+        scheduleDelay: TimeInterval,
+        exportTimeout: TimeInterval,
+        maxQueueSize: Int,
+        maxExportBatchSize: Int,
+        willExportCallback: ((inout [SpanData]) -> Void)?
+    ) {
+        self.spanExporter = spanExporter
+        self.meterProvider = meterProvider
+        self.scheduleDelay = scheduleDelay
+        self.exportTimeout = exportTimeout
+        self.maxQueueSize = maxQueueSize
+        halfMaxQueueSize = maxQueueSize >> 1
+        self.maxExportBatchSize = maxExportBatchSize
+        self.willExportCallback = willExportCallback
+        queue = OperationQueue()
+        queue.name = "BatchWorker Queue"
+        queue.maxConcurrentOperationCount = 1
+        
+        let meter = meterProvider.meterBuilder(name: "io.opentelemetry.sdk.trace").build()
+        self.queueSizeGauge = meter.gaugeBuilder(name: "queueSize")
+            .ofLongs()
+            .setDescription("The number of items queued")
+            .setUnit("1")
+            .buildWithCallback { result in
+                result.record(
+                    value: maxQueueSize,
+                    attributes: [
+                        BatchSpanProcessor.SPAN_PROCESSOR_TYPE_LABEL: .string(BatchSpanProcessor.SPAN_PROCESSOR_TYPE_VALUE)
+                    ]
+                )
+            }
+        
+        self.spanGaugeBuilder = meter.gaugeBuilder(name: "spanSize")
+            .ofLongs()
+        
+        processedSpansCounter = meter.counterBuilder(name: "processedSpans")
+            .setUnit("1")
+            .setDescription("The number of spans processed by the BatchSpanProcessor. [dropped=true if they were dropped due to high throughput]")
+            .build()
+        droppedAttrs = [
+            BatchSpanProcessor.SPAN_PROCESSOR_TYPE_LABEL: .string(BatchSpanProcessor.SPAN_PROCESSOR_TYPE_VALUE),
+            BatchSpanProcessor.SPAN_PROCESSOR_DROPPED_LABEL: .bool(true)
+        ]
+        exportedAttrs = [
+            BatchSpanProcessor.SPAN_PROCESSOR_TYPE_LABEL: .string(BatchSpanProcessor.SPAN_PROCESSOR_TYPE_VALUE),
+            BatchSpanProcessor.SPAN_PROCESSOR_DROPPED_LABEL: .bool(false)
+        ]
+    }
   
   func addSpan(span: ReadableSpan) {
     cond.lock()
     defer { cond.unlock() }
     
     if spanList.count == maxQueueSize {
-      // TODO: Record a counter for dropped spans.
+        processedSpansCounter.add(value: 1, attribute: droppedAttrs)
       return
     }
-    // TODO: Record a gauge for referenced spans.
     spanList.append(span)
+      
+      // Recording span size gauge
+      _ = self.spanGaugeBuilder
+          .buildWithCallback { [count = spanList.count] result in
+              result.record(
+                  value: count,
+                  attributes: [
+                      BatchSpanProcessor.SPAN_PROCESSOR_TYPE_LABEL: .string(BatchSpanProcessor.SPAN_PROCESSOR_TYPE_VALUE)
+                  ]
+              )
+          }
```

Comment:
@bryce-b please take a look

Reply:
yes, I believe you should hang onto the instruments and close any with a callback when the batchSpanProcessor is dealloc'd . (this may need to be done anywhere these are used)